### PR TITLE
Add formatters to devtools testsuite.

### PR DESCRIPTION
**Pull request by epicallan**
`test/Examples/**` contains files that are formatted with respective formaters. 

On the side note, I am not so thrilled with ormolu despite its impressive community adoption. I find its formatting style different from the norm. (twitter thread about the same https://twitter.com/FlavioCorpa/status/1205931199006461954)

~TODO: add Brittany~

I am more inclined to pick Brittany if I had to pick between the three

### CODE REVIEW

File: src/Devtools/Prelude.hs [hunk 1]
```diff
@@ -1,3 +1,9 @@
...
+module Devtools.Prelude
+  ( module Exports
+  , FilePath
+  ) where
 
-import MPrelude as Exports
+import Data.Functor as Exports (fmap)
+import MPrelude     as Exports
+
+type FilePath = String
```
Comment (epicallan) [inline src/Devtools/Prelude.hs]: In another iteration could change to using a better type

**Comment by epicallan** [comment]
I am improving this PR, so it's not ready for merging. I am adding a `FormatProvider` class so that it's easy for libraries users to support any other formatter of their choosing.

I also would like to do some improvements like auto-generating formatter config data instead of asking users to provide it.

**Comment by mbj** [comment]
@epicallan As discussed in channel lets reduce the formatters to `brittany` initially as this is the one we'd use.

File: src/Devtools/Prelude.hs (hunk 1 above)
Comment (dkubb) [reply src/Devtools/Prelude.hs]: The `filepath` package is already a direct dependency on this project. It is also a type alias to `String`, so this line could be replaced with:

```haskell
import System.FilePath (FilePath)
```

File: src/Devtools.hs [hunk 2]
```diff
@@ -1,31 +1,61 @@
...
+  ) where
 
 import Devtools.Prelude
-import System.IO (putStrLn)
 
-import qualified Devtools.Dependencies as Dependencies
-import qualified Devtools.HLint        as HLint
-import qualified Test.Tasty            as Tasty
+import qualified Devtools.Brittany       as Brittany
+import qualified Devtools.Dependencies   as Dependencies
```
Comment (dkubb) [inline src/Devtools.hs]: There is extra whitespace before the `as` here.

File: src/Devtools.hs [hunk 3]
```diff
@@ -1,31 +1,61 @@
...
 
-newtype Config = Config
-  { hlintArguments :: [String]
+data Formatter
+  = Brittany Brittany.Config
+  | Ormolu
+  | Stylish Stylish.Config
+
+data FormatterOptions = FormatterOptions
+  { formatter     :: Formatter
```
Comment (dkubb) [inline src/Devtools.hs]: There is extra whitespace before the `::`.

File: src/Devtools/Brittany.hs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+getBrittanyConfig :: Maybe FilePath  -> IO Brittany.Config
+getBrittanyConfig mFilePath = do
+  mConfig <- runMaybeT $
+    maybe
+      (pure Brittany.staticDefaultConfig)
+      (Brittany.readConfigs mempty . pure)
+      mFilePath
+
+  let errorMsg
+        = "Failed to obtain Brittany Config from provided file"
```
Comment (dkubb) [inline src/Devtools/Brittany.hs]: Did you mean to use more than 2 space indent?

(Sorry for all the formatting comments, I realize these will be 100% eliminated thanks to this PR, and I am thankful for it. If my brain knows there are no whitespace problems, then I'll be able to immediately drop down into more intense reviews)

File: src/Devtools/Stylish.hs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import qualified Data.Text                as Text
+import qualified Data.Text.IO             as Text
+import qualified Language.Haskell.Stylish as Stylish
+import qualified Test.Tasty               as Tasty
+
+testTree :: Stylish.Config ->[FilePath] -> Tasty.TestTree
+testTree config = Tasty.testGroup "Stylish-Haskell" . fmap singleTest
+  where
+    singleTest :: FilePath -> Tasty.TestTree
+    singleTest filePath = goldenTest filePath filePath $ getFormatted config filePath
```
Comment (dkubb) [inline src/Devtools/Stylish.hs]: Maybe it's too "magical" but this could also be written as:

```haskell
singleTest = join goldenTest `ap` getFormatted config
```

My Haskell skills are not as advanced as yours, so I suspect your existing code is better, but I figured I'd still submit it for consideration just in case.

File: test/Examples/Brittany.hs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  , Formatter(..)
+  , FormatterOptions(..)
+  , defaultConfig
+  , defaultMain
+  , main
+  , testTree
+  )
+where
+
+import           Devtools.Prelude
```
Comment (dkubb) [inline test/Examples/Brittany.hs:12]: There is extra whitespace after the `import`.

File: test/Examples/Brittany.hs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import qualified Devtools.Stylish              as Stylish
+import qualified Test.Tasty                    as Tasty
+
+data Formatter
+  = Brittany Brittany.Config
+  | Ormolu
+  | Stylish Stylish.Config
+
+data FormatterOptions = FormatterOptions
+  { formatter     :: Formatter
```
Comment (dkubb) [inline test/Examples/Brittany.hs:27]: There is extra whitespace before the `::`.

File: test/Examples/Ormolu.hs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    main,
+    testTree,
+  )
+where
+
+import qualified Devtools.Brittany as Brittany
+import qualified Devtools.Dependencies as Dependencies
+import qualified Devtools.HLint as HLint
+import qualified Devtools.Ormolu as Ormolu
+import Devtools.Prelude
```
Comment (dkubb) [inline test/Examples/Ormolu.hs:16]: Did you mean to intermix this `import` with the other qualified imports?

File: test/Examples/Ormolu.hs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import qualified Devtools.Stylish as Stylish
+import qualified Test.Tasty as Tasty
+
+data Formatter
+  = Brittany Brittany.Config
+  | Ormolu
+  | Stylish Stylish.Config
+
+data FormatterOptions = FormatterOptions
+  { formatter :: Formatter,
```
Comment (dkubb) [inline test/Examples/Ormolu.hs:26]: The trailing comma should be moved to the next line.

File: test/Examples/Stylish.hs [hunk 10]
```diff
@@ -0,0 +1,60 @@
+module Devtools
```
Comment (dkubb) [inline test/Examples/Stylish.hs]: I've gotta say, Stylish looks a bit more like our "common style" than the others.

Also I like how it does not bother to align the `as` in imports, this will cause a lot of churn (although I do it when I am formatting by hand).

It also eliminates unnecessary whitespace from lots of places instead of leading it "just in case" -- although the other tools may be leaving the whitespace to minimize churn too.

File: test/Test.hs [hunk 11]
```diff
@@ -1,6 +1,44 @@
...
+          }
+
+  let stylishFormatter =
+        Devtools.FormatterOptions
+          { formatter = Devtools.Stylish stylishConfig
+          , files = ["test/Examples/Stylish.hs"]
+          }
+
+  let brittanyFormatter =
+        Devtools.FormatterOptions
```
Comment (dkubb) [inline test/Test.hs]: Did you mean to use more than a 2 space indent here (Same with the other two groups above)?

File: test/Examples/Stylish.hs (hunk 10 above)
Comment (mbj) [reply test/Examples/Stylish.hs]: Note that this is pre-configuration. I suspect that the others can also be "bend" more to our style.

Other than that: I predict once formatting is automatic we do not "see" style that much anymore.

File: test/Test.hs [hunk 12]
```diff
@@ -1,6 +1,44 @@
...
+          { formatter = Devtools.Brittany brittanyConfig
+          , files = ["test/Examples/Brittany.hs"]
```
Comment (dkubb) [inline test/Test.hs]: The `=` should be aligned with the line above? (Same with the other two groups above)

File: test/Test.hs [hunk 13]
```diff
@@ -1,6 +1,44 @@
...
+    testGroup
+      "devtools tests"
+      [ Devtools.testTree $ withFormatterOptions ormoluFormatter
+      , Devtools.testTree $ withFormatterOptions stylishFormatter
+      , Devtools.testTree $ withFormatterOptions brittanyFormatter
+      ]
+  where
+    withFormatterOptions :: Devtools.FormatterOptions -> Devtools.Config
+    withFormatterOptions options = Devtools.Config
+      { hlintArguments = []
```
Comment (dkubb) [inline test/Test.hs]: The `=` should be aligned with the line below.

**Comment by dkubb** [comment]
@epicallan If one formatter is chosen, what do you think about two extra commits in this PR:

1. Fix code to be reformatted with "X"
2. Change Github Actions to check code formatting

For (2), we could reformat the existing code and check to make sure it matches whatever is currently checked in.

I am also fine with doing these things in a follow-up.

File: src/Devtools/Prelude.hs (hunk 1 above)
Comment (epicallan) [reply src/Devtools/Prelude.hs]: Good catch.However I am planning on using the path types in https://hackage.haskell.org/package/path in final PR.

Comment (dkubb) [reply src/Devtools/Prelude.hs]: @epicallan Ahh this is nice. I like their motivation section:

https://github.com/commercialhaskell/path#motivation

Now that I think about it, I am going to purge usage of `filepath` from my own projects.

File: test/Examples/Stylish.hs (hunk 10 above)
Comment (epicallan) [reply test/Examples/Stylish.hs]: Stylish style configuration scope is mainly around module imports so its very narrow. Brittany has more opinions than we can bend to our style.

File: test/Test.hs (hunk 11 above)
Comment (epicallan) [reply test/Test.hs]: They are 2 spaces from the start of `brittany`. So, I meant to use 2 spaces though I can see how it can be confused by 4. BTW shifting Devtool 2 spaces left to align with `brittany` won't parse correctly with in that `let` bind

File: test/Test.hs (hunk 12 above)
Comment (epicallan) [reply test/Test.hs]: Yeah, good catch.

File: src/Devtools/Stylish.hs (hunk 5 above)
Comment (epicallan) [reply src/Devtools/Stylish.hs]: I like this but its quite magical. I am not sure its worth the extra brain cycles it invokes. I feel inclined to leave as is.

**Comment by epicallan** [comment]
>  If one formatter is chosen, what do you think about two extra commits in this PR:

@dkubb 

I am fine with those 2 extra commits, will have them in final PR. Thanks for the review.

File: brittany.yaml [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  lconfig_cols: 80
+  lconfig_indentAmount: 2
+conf_forward:
+  options_ghc:
+  - -XLambdaCase
+  - -XMultiWayIf
+  - -XGADTs
+  - -XPatternGuards
+  - -XViewPatterns
+  - -XTupleSections
```
Comment (epicallan) [inline brittany.yaml]: We don't necessarily need to list all the extensions. Some are already natively supported. However, I need to test this current config on other repos to update it. 

In a future followup PR. I plan to extract the extensions from the cabal file of each formatted project using https://github.com/kowainik/extensions

**Comment by epicallan** [comment]
> Change Github Actions to check code formatting

This is already supported by virtual of having the formatting test as part of the current pre-existing test function

@dkubb

File: src/Devtools/Brittany.hs [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+showBrittanyError = \case
+  ErrorInput         str -> str
+  ErrorUnusedComment str -> str
+  LayoutWarning      str -> str
+  ErrorUnknownNode str _ -> str
+  ErrorMacroConfig str _ -> "When parsing inline config: " <> str
+  ErrorOutputCheck       -> "Output is not syntactically valid."
+
+defaultConfig :: Brittany.Config
+defaultConfig = Brittany.staticDefaultConfig
```
Comment (epicallan) [inline src/Devtools/Brittany.hs:49]: Just FYI @mbj 

The data-files way of bundling a YAML configuration with the project that I had before was based on the guide below. The downside to it is the following:

-  The data file had to be in a directory within the project such as `data` while using `stack` as a build tool. Attempting having them at the root would lead to confusing errors

- The data files direction works well when the project is going to be used as an executable. If you attempt to use it as a library import; the data files wouldn't be accessible. You would need to do some more work to retrieve them.


http://neilmitchell.blogspot.com/2008/02/adding-data-files-using-cabal.html

File: src/Devtools/Brittany.hs [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  { Brittany._conf_forward = Brittany.ForwardOptions syntaxExtensions
+  }
+ where
+  additionalExtensions :: [String]
+  additionalExtensions =
+    ["-XExplicitNamespaces", "-XPatternSynonyms", "-XNumericUnderscores"]
```
Comment (epicallan) [inline src/Devtools/Brittany.hs:61]: Will automatically retrieve extensions in this follow up https://github.com/mbj/devtools-hs/pull/22

File: src/Devtools.hs [hunk 17]
```diff
@@ -1,31 +1,59 @@
...
+  , formatterOptions :: FormatterOptions
   }
 
-defaultConfig :: Config
-defaultConfig = Config
-  { hlintArguments = []
+data FormatterOptions = FormatterOptions
+  { configFile    :: Maybe Path.RelFile
+  , directories   :: NE.NonEmpty Path.RelDir
+  , filesToIgnore :: [Path.RelFile]
```
Comment (epicallan) [inline src/Devtools.hs:30]: There are cases where even a `formatters inline ignore annotations/comments` don't help in dealing with un-supported syntax or edge cases that doesn't lead to module breakage. In such cases we bypass formatting such files entirely

E.g GADTs formatting issue in Brittany https://github.com/lspitzner/brittany/issues/242

**Comment by dkubb** [comment]
@epicallan So the final commit adds Brittany support *and* Refactors the code to match it's new style?

**Comment by epicallan** [comment]
> So the final commit adds Brittany support and Refactors the code to match its new style?

Yes, it adds it to the list of tests run and does the refactoring @dkubb 

update
---

I had actually initially split the commit that adds Brittany and the one that adds formatting into 2. Must have merged them into 2 by mistake. 

Going to update PR and re-separate them

**Comment by epicallan** [comment]
>So the final commit adds Brittany support and Refactors the code to match its new style?


@dkubb I have split the previously one commit into 2.

**Comment by dkubb** [comment]
@epicallan It's up to you, but sometimes when I have similar kinds of changes I will run the formatter (or linter, etc), and commit the fixes it recommends. Then I will do a follow-up commit that installs the formatter/linter so that it runs against the repo.

I think it's ok to commit a formatting fix before the commit that *enforces* it.

File: test/Test.hs [hunk 18]
```diff
@@ -1,6 +1,23 @@
...
 
+import qualified Data.List                     as List
+import qualified Data.Text                     as Text
 import qualified Devtools
+import qualified System.Path                   as Path
 
 main :: IO ()
-main = Devtools.main Devtools.defaultConfig
+main = do
+  files <- List.sort <$> findHaskellFiles [Path.relDir "src"]
```
Comment (dkubb) [inline test/Test.hs:13]: @epicallan Did you make it so the existing `src/**/*.hs` files became part of the golden test? If so, this is pretty cool.

Comment (epicallan) [reply test/Test.hs:13]: Yes, Thanks.

Thinking about this some more I realise I could have used this glob package `http://hackage.haskell.org/package/Glob` for the functionality offered by `findHaskellFiles`

I will experiment replacing my custom code with the glob package, in a followup PR.

File: src/Devtools/Files.hs [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+findFilesRecursive :: Path.RelDir -> IO [Path.RelFile]
+findFilesRecursive path = do
+  filePath <- maybe (fail "File doesn't not exist") pure $ Path.fileFromDir path
+
+  Path.doesFileExist filePath >>= \case
+    True -> pure [filePath]
+    _    -> Path.doesDirectoryExist path >>= \case
+      True -> findFilesRecursive' path
+        <&> List.filter ((== ".hs") . Path.takeExtension)
```
Comment (mbj) [inline src/Devtools/Files.hs:24]: As a followup we may want to include `hs-boot` files here.

File: stack-8.8.yaml [hunk 20]
```diff
@@ -4,6 +4,8 @@ extra-deps:
 - source-constraints-0.0.1@sha256:abdda6ca3d942c81d61ed4374dbc0eee21e9c074da3f17b25d434c2d55df2acc,2979
 - tasty-1.3.1@sha256:01e35c97f7ee5ccbc28f21debea02a38cd010d53b4c3087f5677c5d06617a507,2520
 - tasty-mgolden-0.0.1@sha256:07fadb592cde353c386c5f38cff75539a5675e9350776ed6f4f0a70247427a78,4214
+- git: https://github.com/lspitzner/brittany.git
+  commit: 55f84c329414814c718a81f590fece76c3f91ad1
```
Comment (mbj) [inline stack-8.8.yaml]: Too bad there is no release for us, this forces all devtools users (?) to carry the brittany reference.

**Comment by mbj** [comment]
@epicallan One core property of formatters should be `parse(original_source) == parse(generate(parse(original_source)))` do you know if brittany is tested against this property? I know first hand the GHC ast is not nicely `Eq`-able. So potentially they (or we) have to test against the weaker:

```
generate(parse(original_source)) == generate(parse(generate(parse(original_source))))
```

Which at least guarantees the formatting is stable and on the 2nd round trip does not loose information.

Alternatively we may be able to explore the `Data` instances of the GHC AST to generate an external `Eq` instance, or use `Outputtable` to get a `String` we can test against.

Ideally doing all this while avoiding re-parse operations. Note this is not a flag but a thought. It may be the case we have to go a bit deeper into brittanies internals to get a hold of the first parse nodes to test this property without redundant computation.

I know first hand on my work on [unparser](https://github.com/mbj/unparser) that this round trip property easily fails in edge cases and is a good source of edge case bugs to avoid.

**Comment by epicallan** [comment]
@mbj  they do have an idempotency test for the case you describe above. It tests `parse(original_source) == parse(generate(parse(original_source)))` so I think that is sufficient for now for our case. We can revisit this if we realise it doesn't hold for some cases.

https://github.com/lspitzner/brittany/tree/master/src-idemtests

File: stack-8.8.yaml (hunk 20 above)
Comment (epicallan) [reply stack-8.8.yaml]: This particular release has patches that allow Brittany to work with code that has pattern synonyms among a few others; we could alternatively fall back to the official release and just tell Brittany not to bother formatting that kind of code through an inline comment directive.

File: src/Devtools/Files.hs (hunk 19 above)
Comment (epicallan) [reply src/Devtools/Files.hs:24]: Yes, going to.

File: stack-8.8.yaml (hunk 20 above)
Comment (epicallan) [reply stack-8.8.yaml]: I have removed this brittany reference.
